designs/s3/20200210-design-vscode-s3.md

```diff
@@ -0,0 +1,146 @@
+# UX: S3 explorer
+
+* **Type**: UX Design
+* **Status**: Proposal
+
+## Overview
+
+This document proposes a user experience for interacting with AWS S3 from the
+AWS Toolkit ("the Toolkit") for Visual Studio Code ("VSCode").
+
+Some of the features listed below are vague; they are implemented using VSCode
+or platform conventions analogous to the AWS web console.
+
+## Experience
+
+Each region of the Toolkit explorer tree includes an S3 node (see mockup
+below). The toolkit pulls a list of names and properties of S3 objects in the
+currently active profile and presents this data as a tree in the toolkit
+explorer. This is typical of existing toolkit features such as "Schemas" and
+"CDK".
+
+- When a region is expanded, make an eager request to get the count of S3
```

Comment:
This is potentially expensive since region is not exposed on the `list-buckets` API - in order to determine the region for a bucket you need to make a `head-bucket` call for every bucket in the list. 

This more-or-less mandates you introduce caching of resources. When you switch regions (or add another region) you don't want to have to enumerate the whole list again.

Reply:
All toolkits cache resources (whether they realize it or not), because the server does not have "push" access to the client.